### ISSUE TITLE
Java Problem: The inherited method I.m() cannot hide the public abstract method in J

### ISSUE THREAD

**Issue by khatchad** [khatchad/Migrate-Skeletal-Implementation-to-Interface-Refactoring#115]
Consider the following example:

## Before

```java
package p;
class I {
	void m() { //M''
	}
}

interface J {
	void m(); //M'
}

abstract class A extends I implements J {
	@Override
	public void m() { //M
	}
}
```

## After

```java
package p;
class I {
	void m() { //M''
	}
}

interface J {
	default void m() { //M
	}
}

abstract class A extends I implements J {
}
```

In "after," we now have this problem:

```
The inherited method I.m() cannot hide the public abstract method in J	.
```

This is related to the "class wins" rule, I believe. Basically, we have that `A` "inherits" `m()` from two different locations, namely, `I` and `J`. However, javac picks `I.m()` because the class wins. But, by doing this. `A` is also implementing `J.m()`.  In migrating `A.m()` to a default method of `J`, `A` now inherits `I.m()`, which isn't `public`.

Actually, there is a *much* larger issue here in that the run time target of `C.m()` is changing. So, it's not just the visibility.

This is a very interesting case and I think it's related to the overriding problem in [tip2003].

**Comment by khatchad** [comment]
@mdarefin, this may be one also worth looking into.

**Comment by mdarefin** [comment]
@khatchad looks interesting problem. I will work on it.

**Comment by khatchad** [comment]
This is related to #111 because `A.m()` both implements `I.m()` and overrides `C.m()`. This should probably be marked as a duplicate of #111.

What do you think @mdarefin?

**Comment by mdarefin** [comment]
@khatchad they look similar problem. When I run the similar type, I was keep getting null pointer at main class

**Comment by khatchad** [comment]
@mdarefin Do you have a stack trace?

**Comment by mdarefin** [comment]
on, I compile at the local machine

**Comment by khatchad** [comment]
Well, @mdarefin, if you're getting a `NullPointerException`, you can call `printStackTrace()` on it to see the stack trace. Please paste it here.

**Comment by mdarefin** [comment]
when class 'A' inheriting method m() from interface I and class C. is there any way we could specify

**Comment by khatchad** [comment]
@mdarefin What do you mean by "specify?"

**Comment by mdarefin** [comment]
like the class A getting specific from the interface or from the extended class, if we can define where the method from this problem could be solved.

**Comment by khatchad** [comment]
@mdarefin You mean like the example above?

**Comment by mdarefin** [comment]
yes.

**Comment by khatchad** [comment]
Do you mean you want to run it as a test case?

**Comment by khatchad** [comment]
It would be interesting to see what the error message is.

**Comment by khatchad** [comment]
By the way, the console output is quite long sometimes. I normally search for `edu.cuny` a few times until I get to the error messages.

**Comment by mdarefin** [comment]
wouldn't that be easier? we could take a loot at the type hierarchy of the method

**Comment by khatchad** [comment]
You could also run it in "user mode." The error messages should show up in the refactoring dialog.

**Comment by khatchad** [comment]
Yes, I think it would be best to run a test case using the above example (I'm not sure if we already have one).

**Comment by mdarefin** [comment]
ok, sounds good :)
